### PR TITLE
Refactor getCartDataJs

### PR DESCRIPTION
**Pull request by YevhenShBolt**

### CODE REVIEW

**Comment by daisy1754** [comment]
👍 I like the direction

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 1]
```diff
@@ -189,185 +190,263 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
-            }
+    /**
+     * Generate cart data
+     *
+     * @param $orderCreationResponse
+     * @return array
+     */
+    public function buildCartData($orderCreationResponse)
+    {
+        $authCapture = (Mage::getStoreConfigFlag('payment/boltpay/auto_capture') === self::AUTO_CAPTURE_ENABLED);
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: why not simply `Mage::getStoreConfigFlag('payment/boltpay/auto_capture')` ?

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 2]
```diff
@@ -189,185 +190,263 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+
+        //////////////////////////////////////////////////////////////////////////
+        // Generate JSON cart and hints objects for the javascript returned below.
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: I think we can remove this comment now

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 3]
```diff
@@ -189,185 +190,263 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+     * @return string
+     */
+    public function buildBoltCheckoutJavascript($checkoutType, $immutableQuote, $hintData, $cartData)
+    {
+        /* @var Bolt_Boltpay_Helper_Api $boltHelper */
+        $boltHelper = Mage::helper('boltpay');
+
+        $jsonCart = json_encode($cartData);
+        $jsonHints = '{}';
+        if (sizeof($hintData) != 0) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: comment line below suggests we don't need this if statement - if $hintData is empty $jsonHints will be an empty object.

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 4]
```diff
@@ -189,185 +190,263 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+    /**
+     * Generate BoltCheckout Javascript for output.
+     *
+     * @param $checkoutType
+     * @param $immutableQuote
+     * @param $hintData
+     * @param $cartData
+     * @return string
+     */
+    public function buildBoltCheckoutJavascript($checkoutType, $immutableQuote, $hintData, $cartData)
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: I prefer not passing $immutableQuote here. Maybe pass @immutableQuoteId instead?

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php (hunk 1 above)
Comment (YevhenShBolt) [reply app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: agree

Comment (YevhenShBolt) [reply app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: Previously it was 1 or 0. and it cause the different types. I will refactor it now.

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 5]
```diff
@@ -189,185 +195,268 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+     * @param string $successCustom
+     * @param $checkoutType
+     * @return string
+     */
+    public function buildOnSuccessCallback($successCustom = '', $checkoutType)
+    {
+        $saveOrderUrl = $this->getUrl('boltpay/order/save');
+
+        if ($checkoutType === self::CHECKOUT_TYPE_ADMIN) {
+            $onSuccessCallback = "function(transaction, callback) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: you don't need `$onSuccessCallback`. you can just return.

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 6]
```diff
@@ -189,185 +195,268 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+                document.getElementById('edit_form').appendChild(input);
+
+                // order and order.submit should exist for admin
+                if ((typeof order !== 'undefined' ) && (typeof order.submit === 'function')) {
+                    order_completed = true;
+                    callback();
+                }
+            }";
+        } else {
+            $onSuccessCallback = "function(transaction, callback) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: ditto

File: tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        $cartData = array(
+            'authcapture' => $autoCapture,
+            'orderToken' => md5('bolt'),
+            'error' => $apiErrorMessage
+        );
+
+        $this->assertEquals($cartData, $result, 'Something wrong with testBuildCartDataWithApiError()');
+    }
+
+    public function testIsBoltOnlyPayment()
```
Comment (daisy1754) [inline tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php]: please remove unimplemented tests

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php (hunk 5 above)
Comment (YevhenShBolt) [reply app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: @daisy1754  Are you sure? Because it was there previously.

Comment (daisy1754) [reply app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: @YevhenShBolt what I mean by my comment is you can write:

```
if ($checkoutType === self::CHECKOUT_TYPE_ADMIN) {
    return "function(transaction, callback) {
        $successCustom
        var input = document.createElement('input');
        input.setAttribute('type', 'hidden');
        input.setAttribute('name', 'bolt_reference');
        input.setAttribute('value', transaction.reference);
        document.getElementById('edit_form').appendChild(input);
        // order and order.submit should exist for admin
        if ((typeof order !== 'undefined' ) && (typeof order.submit === 'function')) {
            order_completed = true;
            callback();
        }
    }";
} 
```

No need for variable `$onSuccessCallback`

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 8]
```diff
@@ -40,9 +43,9 @@ class Bolt_Boltpay_Block_Checkout_Boltpay extends Mage_Checkout_Block_Onepage_Re
     const JS_URL_PROD = 'https://connect.bolt.com';
 
     /**
-     * @var int flag that represents if the capture is automatically done on authentication
+     * @var bool flag that represents if the capture is automatically done on authentication
      */
-    const AUTO_CAPTURE_ENABLED = 1;
+    const AUTO_CAPTURE_ENABLED = true;
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: I believe this is no longer needed.

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 9]
```diff
@@ -189,185 +190,266 @@ public function getCartDataJs($checkoutType = 'multi-page')
...
+     * @param $closeCustom
+     * @param $checkoutType
+     * @return string
+     */
+    public function buildOnCloseCallback($closeCustom, $checkoutType)
+    {
+        $successUrl = $this->getUrl(Mage::getStoreConfig('payment/boltpay/successpage'));
+
+        if ($checkoutType === self::CHECKOUT_TYPE_ADMIN) {
+            $onClose = "if (order_completed && (typeof order !== 'undefined' ) && (typeof order.submit === 'function')) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: no need to define variable; just return here.

File: tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+        $result = $this->currentMock->getTrackJsUrl();
+
+        $this->assertEquals($expect, $result);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function testGetCssSuffix()
```
Comment (daisy1754) [inline tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php]: IMO we don't need unit test for oneline getter just returns constant. 
Maintenance burden would be more than benefit.

File: app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php [hunk 11]
```diff
@@ -523,7 +605,7 @@ function getAdditionalCSS()
      * Returns the Bolt Button Theme from configuration.
      * @return string
      */
-    function getTheme()
+    public function getTheme()
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php]: no need to make public

File: tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+            'error' => $apiErrorMessage
+        );
+
+        $this->assertEquals($cartData, $result, 'Something wrong with testBuildCartDataWithApiError()');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function testIsTestMode()
```
Comment (daisy1754) [inline tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php]: IMO we don't need test for isTestMode method since it does not has any logic

**Comment by YevhenShBolt** [comment]
I made rebase

**Pull request merged**